### PR TITLE
Surface graphics device creation failure on 'pc-app' instead of rejecting

### LINKED ISSUE #308: `<pc-app>` never becomes ready when a graphics device cannot be created
## Summary

When a graphics device cannot be created — a browser with WebGL disabled, a blocklisted GPU, a hardened enterprise profile — `<pc-app>` produces an **unhandled promise rejection** and never fires `ready`. There is no error surfaced to the page and no fallback, so the app simply never starts.

The engine's `createGraphicsDevice` is documented as falling back through its candidate device types and always appends `'null'` as a final candidate, so the expectation is a graceful degrade. That fallback does not work.

## Root cause

`createGraphicsDevice` invokes each attempt as the **argument** to `Promise.resolve(...)`:

```js
Promise.resolve(deviceCreateFuncs[attempt++]())
    .then((device) => { ... })
    .catch((err) => { console.log(err); next(); });
```

`deviceCreateFuncs[attempt++]()` is evaluated *before* `Promise.resolve` is called and before `.catch` is attached. `new WebglGraphicsDevice(canvas, options)` throws **synchronously** when `canvas.getContext('webgl2')` returns `null`, so the throw escapes the per-attempt `.catch()` entirely. Instead of advancing to the next candidate, the outer promise rejects.

In this repo, `AppElement.connectedCallback` awaits that promise:

https://github.com/playcanvas/web-components/blob/main/src/app.ts#L162

Since `connectedCallback` is `async` and nothing catches, the rejection becomes an unhandled rejection and the rest of boot — `app.init`, hierarchy construction, `preload`, `start`, `_onReady` — never runs.

`backend="webgpu"` degrades to the same path: the WebGPU candidate is only enqueued when `window?.navigator?.gpu` is truthy, so on a machine without WebGPU the first *enqueued* attempt is still WebGL2.

## Reproduction

Verified in jsdom, where `getContext` returns `null`, which is the same condition as a browser with WebGL disabled:

```html
<pc-app backend="webgl2"></pc-app>
```

Result: `Error: WebGL not supported` as an unhandled rejection; `appElement.ready()` never settles; `appElement.app` stays `null`.

With `backend="null"` the same markup boots normally, which is why the test suite's `bootApp()` helper currently *requires* `backend="null"` rather than relying on the documented fallback.

## Suggested fix

Two parts, independently useful:

1. **Upstream (playcanvas/engine)** — defer the attempt so a synchronous throw is captured:
   ```js
   Promise.resolve().then(() => deviceCreateFuncs[attempt++]())
   ```
   This makes the existing `.catch(() => next())` behave as intended and restores the documented fallback for every consumer.

2. **Here** — wrap the `await createGraphicsDevice(...)` in `src/app.ts` so a total failure surfaces as something actionable (a `console.error` naming the attempted backends, and/or an `error` event on the element) instead of a silent never-ready element. A user-visible failure mode is much better than an app that just doesn't appear.

Fixing (1) would also let the test helper drop its `backend="null"` requirement.

## Notes

Found while building the test suite (#307). A regression test is planned for the integration tier; it will reference this issue.

### PR DESCRIPTION
**Pull request by willeastcott**
Fixes #308

## What

When no graphics device can be created (a browser with WebGL disabled, a blocklisted GPU, a hardened enterprise profile), `<pc-app>` previously produced an unhandled promise rejection from its async `connectedCallback` and never fired `ready` - a blank page with a loading bar stuck at zero and no signal to the page.

It now:

- catches the failure and returns the element to its pre-boot state (canvas and loading bar removed),
- logs a `console.error` naming the requested backends, e.g. `pc-app failed to create a graphics device (webgpu, webgl2) - WebGL not supported`, and
- dispatches a non-bubbling `ErrorEvent` named `error` on the element, with the message and the underlying failure in `error` - so a page can show a fallback UI.

## The readiness contract (the 1.0.0 lock-in)

Readiness deliberately **stays pending** on failure. A device-less `<pc-app>` has nothing its readiness would announce - no app, no entity hierarchy - and its descendants can never become ready either, so it joins the documented never-settles cases (`whenReady` docs updated). This also preserves the library-wide invariant that no `ready()` promise ever rejects, and keeps `sky.ts`/`particlesystem-component.ts` (which dereference `.app` after awaiting readiness) safe. This intentionally differs from `pc-model`, where a failed *asset load* on a successfully booted element settles readiness - here the element cannot finish initializing at all.

Removing and re-inserting the element retries the boot with its current attributes (teardown already unlocks the boot-time options), which is the documented recovery path.

## Notes

- The requested-backend list is snapshotted before the `createGraphicsDevice` call because the engine appends its `null`-device fallback to the caller''s `deviceTypes` array in place.
- The CEM validator''s inheritance-leak canary (pc-app must not inherit pc-asset''s events) now pins only `load`, since `error` is legitimately pc-app''s own; a positive check for the new event is added.
- The engine-side half of #308 (its device-type fallback is broken by a synchronous throw escaping `Promise.resolve(fn())`) still stands and is worth filing upstream; this change makes `<pc-app>` robust either way.

## Verification

- 4 new integration tests: error event + never-ready, requested-backend naming, element left empty, and recovery by re-insertion with a working backend. The suite''s guard also asserts no unhandled rejection escapes.
- Full suite: 612 tests pass; type-check, lint, prettier and `cem` validation clean.
- Verified end-to-end against the built `dist/pwc.mjs` in a real browser with `getContext` stubbed to return `null` (the same observable condition as WebGL disabled): error event fired, zero unhandled rejections, element left empty, readiness never settled.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/app.ts [hunk 1]
```diff
@@ -86,6 +86,12 @@ const pointerEventTypes = ['pointermove', 'pointerdown', 'pointerup', 'pointeren
  * `total` are asset counts, not bytes, and an asset that fails to load still counts as loaded.
  * Fired at least once per boot, and the final event always has `loaded` equal to `total`. Does
  * not bubble.
+ *
+ * @fires {ErrorEvent} error - Fired when the application cannot boot because no graphics device
+ * could be created (for example, a browser with WebGL disabled). `message` names the backends
+ * that were attempted and `error` holds the underlying failure. The element never becomes ready
```
[the remark this answers is not in the document]
Comment (willeastcott) [reply src/app.ts]: Fixed in 9526244 - the JSDoc now says 'requested backends', matching what the message reports. 'Requested' is the honest term here: the engine silently skips candidates it cannot enqueue (e.g. webgpu without navigator.gpu) and mutates the list in place, so what was actually attempted is not observable from this side.

File: src/app.ts [hunk 2]
```diff
@@ -241,14 +247,46 @@ class AppElement extends AsyncElement {
...
+                this.removeChild(this._canvas);
+            }
+            this._canvas = null;
+            this._bar?.destroy();
+            this._bar = null;
+
+            const reason = error instanceof Error ? error.message : String(error);
+            const message = `pc-app failed to create a graphics device (${requested}) - ${reason}`;
+            console.error(message);
+            this.dispatchEvent(new ErrorEvent('error', { message, error }));
```
[the remark this answers is not in the document]
Comment (willeastcott) [reply src/app.ts:287]: Fixed in 9526244 - the underlying error is now passed as a second argument to console.error, so the stack and any device-specific fields survive into devtools.
